gym_madras.py

```diff
@@ -0,0 +1,160 @@
+                        # Open new UDP in vtorcs
+                        self.client.MAX_STEPS = self.CLIENT_MAX_STEPS
+                        self.client.get_servers_input(step=0)
+                        # Get the initial input from torcs
+                        raw_ob = self.client.S.d
+                        # Get the current full-observation from torcs
+                        self.ob = self.make_observation(raw_ob)
+                    except:
+                        pass
+
+        self.distance_traversed = 0
+        s_t = np.hstack((self.ob.angle, self.ob.track, self.ob.trackPos,
+                        self.ob.speedX, self.ob.speedY, self.ob.speedZ,
+                        self.ob.wheelSpinVel / 100.0, self.ob.rpm))
+
+        return s_t
+
+    def step(self, desire):
+        """Step Function."""
+        r_t = 0
+
+        for PID_step in range(self.PID_latency):
+                # Implement the desired trackpos and velocity using PID
+            if self.pid_assist:
+                self.accel_PID.update_error((desire[1] - self.prev_vel))
+                self.steer_PID.update_error((-(self.prev_lane - desire[0]) /
+                                            10 + self.prev_angle))
+                if self.accel_PID.output() < 0.0:
+                    brake = 1
+                else:
+                    brake = 0
+                a_t = np.asarray([self.steer_PID.output(),
+                                 self.accel_PID.output(), brake])
+            else:
+                a_t = desire
+            try:
+                self.ob, r, done, info = TorcsEnv.step(self, PID_step,
+                                                       self.client, a_t,
+                                                       self.early_stop)
+            except Exception as e:
+                print(("Exception caught at port " + str(e)))
+                self.ob = None
+                while self.ob is None:
+                    try:
+                        self.client = snakeoil3.Client(p=self.port,
+                                                       vision=self.vision)
+                        # Open new UDP in vtorcs
+                        self.client.MAX_STEPS = self.CLIENT_MAX_STEPS
+                        self.client.get_servers_input(0)
+                        # Get the initial input from torcs
+                        raw_ob = self.client.S.d
+                        # Get the current full-observation from torcs
+                        self.ob = self.make_observation(raw_ob)
+                    except:
+                        pass
+                    continue
+            self.prev_vel = self.ob.speedX
+            self.prev_angle = self.ob.angle
+            self.prev_lane = self.ob.trackPos
+            if (math.isnan(r)):
+                r = 0.0
+            r_t += r  # accumulate rewards over all the time steps
+            # self.distance_traversed+=self.ob.speedX*np.cos(self.ob.angle)
```

Comment:
Remove this line, if it is not in use. If this could be used in future, leave a comment mentioning the use case. Same for line number 153.